contracts/upgradeable_contracts/BridgeValidators.sol

```diff
@@ -9,89 +9,167 @@ import "../upgradeability/EternalStorage.sol";
     }
 
-    function setRequiredSignatures(uint256 _requiredSignatures) external onlyOwner {
+    function setRequiredSignatures(uint256 _requiredSignatures)
+        external
+        onlyOwner
+    {
         require(validatorCount() >= _requiredSignatures);
         require(_requiredSignatures != 0);
         uintStorage[keccak256(abi.encodePacked("requiredSignatures"))] = _requiredSignatures;
         emit RequiredSignaturesChanged(_requiredSignatures);
     }
 
-    function getBridgeValidatorsInterfacesVersion() public pure returns(uint64 major, uint64 minor, uint64 patch) {
+    function getBridgeValidatorsInterfacesVersion()
+        public
+        pure
+        returns (uint64 major, uint64 minor, uint64 patch)
+    {
         return (2, 0, 0);
     }
 
-    function requiredSignatures() public view returns(uint256) {
+    function requiredSignatures() public view returns (uint256) {
         return uintStorage[keccak256(abi.encodePacked("requiredSignatures"))];
     }
 
-    function validatorCount() public view returns(uint256) {
+    function validatorCount() public view returns (uint256) {
         return uintStorage[keccak256(abi.encodePacked("validatorCount"))];
     }
 
-    function validators(address _validator) public view returns(bool) {
-        return boolStorage[keccak256(abi.encodePacked("validators", _validator))];
-    }
-
-    function isValidator(address _validator) public view returns(bool) {
-        return validators(_validator) == true;
+    function isValidator(address _validator) public view returns (bool) {
+        return _validator != F_ADDR && getNextValidator(_validator) != address(0);
     }
 
-    function isInitialized() public view returns(bool) {
+    function isInitialized() public view returns (bool) {
         return boolStorage[keccak256(abi.encodePacked("isInitialized"))];
     }
 
-    function deployedAtBlock() public view returns(uint256) {
+    function deployedAtBlock() public view returns (uint256) {
         return uintStorage[keccak256("deployedAtBlock")];
     }
 
+    function getValidatorRewardAddress(address _validator) public view returns (address) {
+        return addressStorage[keccak256(abi.encodePacked("validatorsRewards", _validator))];
+    }
+
+    function getNextValidator(address _address) public view returns (address) {
+        return addressStorage[keccak256(abi.encodePacked("validatorsList", _address))];
+    }
+
+    function validatorList() public view returns (address[]) {
+        address [] memory list = new address[](validatorCount());
+        uint256 counter = 0;
+        address nextValidator = getNextValidator(F_ADDR);
+
+        while (nextValidator != F_ADDR) {
+            list[counter] = nextValidator;
+            nextValidator = getNextValidator(nextValidator);
+            counter++;
+        }
+
+        return list;
+    }
+
+    function setValidatorRewardAddress(address _validator, address _reward) internal {
```

Comment:
Should we have a base validator contract without any code related to the reward and another contract like `RewardableValidators` which will inherit functionality of the base validator contract. It will allow to use the base contract in the cases when the subsidized bridge is needed.

Reply:
Yes, that makes sense. Here is the refactor https://github.com/poanetwork/poa-bridge-contracts/pull/130/commits/4e9db7b626ef73a847b019247865d9f0ae1d53d4